### PR TITLE
virtual: factor label selector injection out

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This makes the underlying code more reusable for other types of filters
that virtual workspaces might want to use.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @sttts

### CODE REVIEW

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 1]
```diff
@@ -44,32 +45,38 @@ func newStores(ctx context.Context, gvr schema.GroupVersionResource, dynamicClus
...
-			},
-			DefaultQualifiedResource: resource,
-			CreateStrategy:           strategy,
-			UpdateStrategy:           strategy,
-			DeleteStrategy:           strategy,
-			TableConvertor:           tableConvertor,
-			ResetFieldsStrategy:      strategy,
+		requirements, selectable := labels.SelectorFromSet(labels.Set(labelSelector)).Requirements()
+		if !selectable {
+			// why are you even asking for this store without a selector that can match anything? error
```
Comment (stevekuznetsov) [inline pkg/virtual/framework/forwardingregistry/rest.go:50]: There's no way to bubble an error out of the `apiserver.RestProviderFunc` ... want me to refactor that?

Comment (sttts) [reply pkg/virtual/framework/forwardingregistry/rest.go:50]: if everything is static, a panic is ok'ish.

Comment (sttts) [reply pkg/virtual/framework/forwardingregistry/rest.go:50]: would go with `labels.Selector` as the type you pass in here. Then you can do all the error-prone calls here: https://github.com/kcp-dev/kcp/blob/main/pkg/virtual/syncer/builder/build.go#L116

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 2]
```diff
@@ -44,32 +45,38 @@ func newStores(ctx context.Context, gvr schema.GroupVersionResource, dynamicClus
...
+		}
 
-			resource:                  gvr,
-			dynamicClusterClient:      dynamicClusterClient,
-			patchConflictRetryBackoff: *patchConflictRetryBackoff,
-			labelSelector:             labelSelector,
+		store := &LabelSelectingStore{
```
Comment (sttts) [inline pkg/virtual/framework/forwardingregistry/rest.go:53]: nit (style): 
```
store := &Store {
   ....
}
store := &LabelSelectingStore{store}
```

**Comment by stevekuznetsov** [comment]
@sttts updated - looked at returning an error but it would mean refactoring ... a lot ....

**Comment by stevekuznetsov** [comment]
I don't want to continue waiting for enormous PRs that are WIP. We should work in small, incremental bits. This factoring is useful for my current work, and stands alone.

**Comment by stevekuznetsov** [comment]
We're doing a live GET during the UPDATE flow (!!!!) so this PR is incorrect ... 

/hold

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 3]
```diff
@@ -31,20 +31,22 @@ import (
 	"k8s.io/client-go/util/retry"
 )
 
+type StorageAdapter func(schema.GroupResource, customresource.Store) customresource.Store
```
Comment (sttts) [inline pkg/virtual/framework/forwardingregistry/rest.go:34]: am always falling over "adapter" as a term. What is an adapter? A wrapper has more meaning IMO as it is says something about structure.

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 4]
```diff
@@ -57,26 +59,27 @@ func newStores(ctx context.Context, gvr schema.GroupVersionResource, dynamicClus
...
-			labelSelector:             labelSelector,
-
-			stopWatchesCh: ctx.Done(),
+			stopWatchesCh:             ctx.Done(),
 		}
+		store := adapter(resource, delegate)
+		delegate.getter = store
 
-		statusStore := *store // shallow copy
+		statusDelegate := *delegate // shallow copy
```
Comment (sttts) [inline pkg/virtual/framework/forwardingregistry/rest.go:76]: why not store?

File: pkg/virtual/framework/forwardingregistry/rest.go (hunk 3 above)
Comment (stevekuznetsov) [reply pkg/virtual/framework/forwardingregistry/rest.go:34]: Happy to change it to whatever :)

File: pkg/virtual/framework/forwardingregistry/store.go [hunk 5]
```diff
@@ -76,10 +77,13 @@ type Store struct {
 	dynamicClusterClient      dynamic.ClusterInterface
 	subResources              []string
 	patchConflictRetryBackoff wait.Backoff
-	labelSelector             map[string]string
 
 	// stopWatchesCh closing means that all existing watches are closed.
 	stopWatchesCh <-chan struct{}
+
+	// getter is what we use for self-referential GET calls to allow upstream
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/virtual/framework/forwardingregistry/store.go:78]: I have little hope we can improve this, at least for patch. If you transform object, which we definitely want to do, we have to get and update.

**Comment by stevekuznetsov** [comment]
@sttts updated

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 6]
```diff
@@ -31,20 +31,22 @@ import (
 	"k8s.io/client-go/util/retry"
 )
 
+type StorageWrapper func(schema.GroupResource, customresource.Store) customresource.Store
+
 // NewStorage returns a REST storage that forwards calls to a dynamic client
 func NewStorage(ctx context.Context, resource schema.GroupVersionResource, kind, listKind schema.GroupVersionKind, strategy customresource.CustomResourceStrategy, categories []string, tableConvertor rest.TableConvertor, replicasPathMapping fieldmanager.ResourcePathMappings,
-	dynamicClusterClient dynamic.ClusterInterface, patchConflictRetryBackoff *wait.Backoff, labelSelector map[string]string) customresource.CustomResourceStorage {
-	stores := newStores(ctx, resource, dynamicClusterClient, patchConflictRetryBackoff, labelSelector)
+	dynamicClusterClient dynamic.ClusterInterface, patchConflictRetryBackoff *wait.Backoff, adapter StorageWrapper) customresource.CustomResourceStorage {
```
Comment (sttts) [inline pkg/virtual/framework/forwardingregistry/rest.go:38]:
```suggestion
	dynamicClusterClient dynamic.ClusterInterface, patchConflictRetryBackoff *wait.Backoff, wrapper StorageWrapper) customresource.CustomResourceStorage {
```

**Comment by sttts** [comment]
Just a nit. And tests are red.

**Comment by stevekuznetsov** [comment]
Weird, passes for me locally:

```
$ make test-e2e WHAT=./test/e2e/virtual/syncer
NO_GORUN=1 go test -race -count 1 -p 1 -parallel 1 ./test/e2e/virtual/syncer 
ok  	github.com/kcp-dev/kcp/test/e2e/virtual/syncer	31.987s
```

**Comment by stevekuznetsov** [comment]
Also in parallel:

```
$ make test-e2e WHAT=./test/e2e/virtual/syncer COUNT=5 E2E_PARALLELISM=5
NO_GORUN=1 go test -race -count 5 -p 5 -parallel 5 ./test/e2e/virtual/syncer 
ok  	github.com/kcp-dev/kcp/test/e2e/virtual/syncer	104.371s
```

**Comment by stevekuznetsov** [comment]
@sttts updated

File: test/e2e/virtual/syncer/virtualworkspace_test.go [hunk 7]
```diff
@@ -368,8 +368,38 @@ func TestSyncerVirtualWorkspace(t *testing.T) {
...
+			_, err = unrelatedWorkspaceWorkspaceClient.WildwestV1alpha1().Cowboys("default").Create(ctx, &v1alpha1.Cowboy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "unluckyluke",
+				},
+				Spec: v1alpha1.CowboySpec{
+					Intent: "will never catch joe",
+				},
+			}, metav1.CreateOptions{})
+			require.NoError(t, err)
+
```
Comment (sttts) [inline test/e2e/virtual/syncer/virtualworkspace_test.go:402]: trying to understand how this test change is related to the PR

Comment (stevekuznetsov) [reply test/e2e/virtual/syncer/virtualworkspace_test.go:402]: We need to create some object that does not have the label to make this test fail all by itself, as opposed to only fail when some unrelated test creates such a thing and happens to do it at the right time.

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 8]
```diff
@@ -31,20 +31,22 @@ import (
 	"k8s.io/client-go/util/retry"
 )
 
+type StorageWrapper func(schema.GroupResource, customresource.Store) customresource.Store
```
Comment (ncdc) [inline pkg/virtual/framework/forwardingregistry/rest.go:34]: Can you add godoc?

File: pkg/virtual/syncer/builder/forwarding.go [hunk 9]
```diff
@@ -80,3 +86,73 @@ func provideForwardingRestStorage(ctx context.Context, clusterClient dynamic.Clu
...
+// Get implements rest.Getter
+func (s *LabelSelectingStore) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	obj, err := s.Store.Get(ctx, name, options)
+	if err != nil {
+		return obj, err
+	}
+
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return obj, fmt.Errorf("expected a metav1.Object, got %T", obj)
```
Comment (ncdc) [inline pkg/virtual/syncer/builder/forwarding.go:141]: Should we return nil or obj here?

File: pkg/virtual/framework/forwardingregistry/rest.go [hunk 10]
```diff
@@ -57,26 +59,27 @@ func newStores(ctx context.Context, gvr schema.GroupVersionResource, dynamicClus
...
 
 			resource:                  gvr,
 			dynamicClusterClient:      dynamicClusterClient,
 			patchConflictRetryBackoff: *patchConflictRetryBackoff,
-			labelSelector:             labelSelector,
-
-			stopWatchesCh: ctx.Done(),
+			stopWatchesCh:             ctx.Done(),
 		}
+		store := wrapper(resource, delegate)
```
Comment (ncdc) [inline pkg/virtual/framework/forwardingregistry/rest.go:73]: Should wrapper be optional?

Comment (stevekuznetsov) [reply pkg/virtual/framework/forwardingregistry/rest.go:73]: We don't currently have a use-case for it, so if someone comes up with one we can merge the changes required to support it.

**Comment by stevekuznetsov** [comment]
@ncdc addressed comments

File: pkg/virtual/framework/forwardingregistry/rest.go (hunk 10 above)
Comment (ncdc) [reply pkg/virtual/framework/forwardingregistry/rest.go:75]: The APIExport virtual workspace doesn't need it?

Comment (ncdc) [reply pkg/virtual/framework/forwardingregistry/rest.go:75]: I can put it in my PR if needed

**Review by ncdc — APPROVED** [review]

**Pull request merged**